src/python/pants/backend/javascript/dependency_inference/rules.py

```diff
@@ -173,29 +169,57 @@ def _add_extensions(file_imports: frozenset[str], file_extensions: tuple[str, ..
 async def _determine_import_from_candidates(
     candidates: ParsedJavascriptDependencyCandidate,
     package_candidate_map: NodePackageCandidateMap,
+    tsconfig: TSConfig | None,
     file_extensions: tuple[str, ...],
 ) -> Addresses:
-    paths = await path_globs_to_paths(
-        _add_extensions(
-            candidates.file_imports,
-            file_extensions,
-        )
-    )
-    local_owners = await Get(Owners, OwnersRequest(paths.files))
-
-    owning_targets = await Get(Targets, Addresses(local_owners))
-
-    addresses = Addresses(tgt.address for tgt in owning_targets)
-    if not local_owners:
-        non_path_string_bases = FrozenOrderedSet(
-            non_path_string.partition(os.path.sep)[0]
-            for non_path_string in candidates.package_imports
+    addresses = Addresses(())
+    # 1. handle relative imports
+    if candidates.file_imports:
+        paths = await path_globs_to_paths(
+            _add_extensions(
+                candidates.file_imports,
+                file_extensions,
+            )
         )
-        addresses = Addresses(
-            package_candidate_map[pkg_name]
-            for pkg_name in non_path_string_bases
-            if pkg_name in package_candidate_map
+        local_owners = await Get(Owners, OwnersRequest(paths.files))
+        owning_targets = await Get(Targets, Addresses(local_owners))
+        addresses = Addresses(tgt.address for tgt in owning_targets)
+
+    # 2. handle package imports
+    elif candidates.package_imports:
+
+        # Try prepending the tsconfig root dir to the package import.
+        first_party_packge_imports = frozenset(
+            os.path.join(tsconfig.resolution_root_dir, pkg_import)
+            for pkg_import in candidates.package_imports
+        ) if tsconfig and tsconfig.resolution_root_dir else frozenset()
+        paths = await path_globs_to_paths(
+            _add_extensions(
+                first_party_packge_imports,
+                file_extensions,
+            )
         )
+        local_owners = await Get(Owners, OwnersRequest(paths.files))
```

Comment:
This should be handled in the rust parser. If it isn't, it is a bug.

Reply:
This is also why the if-clause is ill-formed. The parser can produce candidates on either form.